File: fiber-lib/src/rpc_types/invoice.rs
use crate::fiber::config::MIN_TLC_EXPIRY_DELTA;
use crate::fiber::hash_algorithm::HashAlgorithm;
use crate::fiber::serde_utils::{U128Hex, U64Hex};
use crate::fiber::types::{Hash256, Privkey};
use crate::invoice::{CkbInvoice, CkbInvoiceStatus, Currency, InvoiceBuilder, InvoiceStore};
use crate::FiberConfig;
use ckb_jsonrpc_types::Script;
use jsonrpsee::types::error::CALL_EXECUTION_FAILED_CODE;
use jsonrpsee::{core::async_trait, proc_macros::rpc, types::ErrorObjectOwned};
use secp256k1::{PublicKey, Secp256k1, SecretKey};
use serde::{Deserialize, Serialize};
use serde_with::serde_as;
use std::time::Duration;
use tentacle::secio::SecioKeyPair;

/// The parameter struct for generating a new invoice.
#[serde_as]
#[derive(Serialize, Deserialize)]
pub struct NewInvoiceParams {
    /// The amount of the invoice.
    #[serde_as(as = "U128Hex")]
    pub amount: u128,
    /// The description of the invoice.
    pub description: Option<String>,
    /// The currency of the invoice.
    pub currency: Currency,
    /// The payment preimage of the invoice.
    pub payment_preimage: Hash256,
    /// The expiry time of the invoice, in seconds.
    #[serde_as(as = "Option<U64Hex>")]
    pub expiry: Option<u64>,
    /// The fallback address of the invoice.
    pub fallback_address: Option<String>,
    /// The final HTLC timeout of the invoice, in milliseconds.
    #[serde_as(as = "Option<U64Hex>")]
    pub final_expiry_delta: Option<u64>,
    /// The UDT type script of the invoice.
    pub udt_type_script: Option<Script>,
    /// The hash algorithm of the invoice.
    pub hash_algorithm: Option<HashAlgorithm>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct InvoiceResult {
    /// The encoded invoice address.
    pub invoice_address: String,
    /// The invoice.
    pub invoice: CkbInvoice,
}

#[derive(Serialize, Deserialize)]
pub struct ParseInvoiceParams {
    /// The encoded invoice address.
    pub invoice: String,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ParseInvoiceResult {
    /// The invoice.
    pub invoice: CkbInvoice,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct InvoiceParams {
    /// The payment hash of the invoice.
    pub payment_hash: Hash256,
}

/// The status of the invoice.
#[derive(Clone, Serialize, Deserialize)]
pub struct GetInvoiceResult {
    /// The encoded invoice address.
    pub invoice_address: String,
    /// The invoice.
    pub invoice: CkbInvoice,
    /// The invoice status
    pub status: CkbInvoiceStatus,
}

File: fiber-js/src/types/invoice.ts
import { Script } from "./channel";
import { HashAlgorithm, HexString } from "./general"

type Currency = "Fibb" | "Fibt" | "Fibd";
type CkbInvoiceStatus = "Open" | "Cancelled" | "Expired" | "Received" | "Paid";
interface NewInvoiceParams {
    amount: HexString;
    description?: string;
    currency: Currency;
    payment_preimage: HexString;
    expiry?: HexString;
    fallback_address?: string;
    final_expiry_delta?: HexString;
    udt_type_script?: Script;
    hash_algorithm?: HashAlgorithm;
}
type CkbScript = HexString;
type Attribute = { FinalHtlcMinimumExpiryDelta: HexString } |
{ ExpiryTime: HexString } |
{ Description: string } |
{ FallbackAddr: string } |
{ UdtScript: CkbScript } |
{ PayeePublicKey: string } |
{ HashAlgorithm: number } |
{ Feature: HexString };


interface InvoiceData {
    timestamp: HexString;
    payment_hash: HexString;
    attrs: Attribute[];
}
interface CkbInvoice {
    currency: Currency;
    amount?: HexString;
    signature?: string;
    data: InvoiceData;
}

interface InvoiceResult {
    invoice_address: string;
    invoice: CkbInvoice;
}

interface ParseInvoiceParams {
    invoice: string;
}
interface ParseInvoiceResult {
    invoice: CkbInvoice;
}
interface InvoiceParams {
    payment_hash: HexString;
}

interface GetInvoiceResult {
    invoice_address: string;
    invoice: CkbInvoice;
    status: CkbInvoiceStatus;
}

export type {
    NewInvoiceParams,
    InvoiceResult,
    Attribute,
    CkbInvoice,
    CkbInvoiceStatus,
    CkbScript,
    Currency,
    GetInvoiceResult,
    InvoiceData,
    InvoiceParams,
    ParseInvoiceParams,
    ParseInvoiceResult
}
